examples/connect-latency.bpf.c

```diff
@@ -0,0 +1,133 @@
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_core_read.h>
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_LATENCY_SLOT 26
+
+struct connect_start_key_t {
+    u64 pid_tgid;
+};
+
+struct connect_start_val_t {
+    u64 ts;
+    int addrlen;
+    u32 d_ip;    // Destination IPv4 address
+    u16 d_port;  // Destination port number
+};
+
+struct connect_latency_key_t {
+    u32 d_ip;    // Destination IPv4 address
+    u16 d_port;  // Destination port number
+    u64 slot;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 10240);
+    __type(key, struct connect_start_key_t);
+    __type(value, struct connect_start_val_t);
+} connect_start SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, MAX_LATENCY_SLOT + 2);
+    __type(key, struct connect_latency_key_t);
+    __type(value, u64);
+} connect_latency_seconds SEC(".maps");
+
+static inline __u16 ntohs(__u16 value) {
+    return ((value & 0x00FF) << 8) | ((value & 0xFF00) >> 8);
+}
+
+SEC("kprobe/__sys_connect")
```

Comment:
`fentry` is a lot faster:

* https://github.com/cloudflare/ebpf_exporter/tree/master/benchmark

You also get:

* Function arguments in `fexit`, allowing you to key on `addr` rather than pid
* No need for `BPF_CORE_READ` as BTF allows direct reads